sklearn/preprocessing/tests/test_knn_imputation.py

```diff
@@ -0,0 +1,220 @@
+from __future__ import division
+import numpy as np
+
+from sklearn.utils.testing import assert_equal
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_array_almost_equal
+from sklearn.utils.testing import assert_raises
+from sklearn.utils.testing import assert_false
+
+from sklearn.preprocessing.knn_imputation import KNNImputer
+from sklearn.random_projection import sparse_random_matrix
+
+
+def test_knn_imputation_shape():
+    # Verify the shapes of the imputed matrix for different weights and
+    # number of neighbors.
+    n_rows = 10
+    n_cols = 2
+    X = np.random.rand(n_rows, n_cols)
+    X[0, 0] = np.nan
+
+    for weights in ['uniform', 'distance']:
+        for n_neighbors in range(1, 6):
+            imputer = KNNImputer(n_neighbors=n_neighbors, weights=weights)
+            X_imputed = imputer.fit_transform(X)
+            assert_equal(X_imputed.shape, (n_rows, n_cols))
+
+
+def test_knn_imputation_zero():
+    # Test imputation when missing_values == 0
```

Comment:
The structure of this test should be to check that imputation is identical when replacing nans with 0s for a matrix that did not originally have 0s.

Reply:
I'm okay with restricting to NaN. This limits applicability to integer data matrices, but we're already not offering mode imputation by NN so we're not supporting integer outputs anyway.

Reply:
im not sure i follow. wud u mind elaborating a little what limitation would be caused by using only NaN for missing?

Reply:
An array with integer dtype cannot include a NaN:
```
>>> x = np.arange(5)
>>> x[0] = np.nan
ValueError: cannot convert float NaN to integer
```

If we supported mode imputation here then for integer input, the output could be assured to be integer. We don't, so the output is assured to be float, hence the input might as well be float (we don't gain more precision or asymptotic space efficiency by using int), hence we don't need to support choices other than NaN.